torch_geometric/data/remote_backend_utils.py

```diff
@@ -0,0 +1,70 @@
+# This file defines a set of utilities for remote backends (backends that are
+# characterize as Tuple[FeatureStore, GraphStore]).
+from typing import Tuple, Union
+
+from torch_geometric.data.feature_store import FeatureStore
+from torch_geometric.data.graph_store import GraphStore
+from torch_geometric.typing import EdgeType, NodeType
+
+
+def num_nodes(
+    feature_store: FeatureStore,
+    graph_store: GraphStore,
+    query: Union[NodeType, EdgeType],
+) -> Union[int, Tuple[int, int]]:
+    r"""Returns the number of nodes in the node type or the number of source
+    and destination nodes in an edge type by sequentially accessing attributes
+    in the feature and graph stores that reveal this number."""
+    def _matches_edge_type(query: Union[NodeType, EdgeType],
+                           edge_type: EdgeType) -> bool:
+        if isinstance(query, (list, tuple)):  # EdgeType
+            return query == edge_type
+        else:
+            return query == edge_type[0] or query == edge_type[-1]
+
+    def _matches_node_type(query: Union[NodeType, EdgeType],
+                           node_type: NodeType) -> bool:
+        if isinstance(query, (list, tuple)):  # EdgeType
+            return query[0] == node_type or query[-1] == node_type
+        else:
+            return query == node_type
+
+    node_query = isinstance(query, NodeType)
+
+    # 1. Check GraphStore:
+    edge_attrs = graph_store.get_all_edge_attrs()
+    for edge_attr in edge_attrs:
```

Comment:
Any opinion on avoiding the for-loop here? Isn't this an implementation detail how the `GraphStore`/`FeatureStore` save the edge attributes? If it is done as part of some hash map, we should be able to leverage this.

Reply:
Yeah, I think graph store and feature store should expose (optional) methods to obtain a tensorattr/edgeattr from the first member of the corresponding dataclass. But wanted to leave that for another PR.